src/DiscreteSystem.ts

```diff
@@ -101,4 +102,35 @@ export class DiscreteSystem<TShared, TInternal> implements System {
         }
         return state;
     }
+
+    public compareData(data: TData[] | TData) {
+        // are we sharing data between all groups
+        const isSharedData = !Array.isArray(data) || data.length === 1;
+        let sharedData: TData;
+        if (isSharedData) {
+            sharedData = Array.isArray(data) ? data[0] : data;
+        } else {
+            if (data.length !== this._nGroups) {
+                throw new RangeError("Expected data to have same length as groups.");
+            }
+        }
+
+        const result = new SystemDataComparison(this._nGroups, this._nParticles);
+        this.iterateParticles((iGroup: number, iParticle: number) => {
+            const iData = isSharedData ? sharedData : data[iGroup];
+            const state = this._state.getParticle(iGroup, iParticle);
+            const shared = this._shared[iGroup];
+            const internal = this._internal[iGroup];
+            const comparisonValue = this._generator.compareData(
+                this._time,
+                particleStateToArray(state),
+                iData,
+                shared,
+                internal,
+                this._random
+            );
+            result.setValue(iGroup, iParticle, comparisonValue);
```

Comment:
See comment in SystemDataComparison, but I don't see why this is not just writing into a vector/matrix directly in this function